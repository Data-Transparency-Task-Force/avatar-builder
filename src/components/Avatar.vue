<template>
  <!-- should use func(...arguments) notation for for stubs in tests -->
  <svg
    version="1.1"
    id="Layer_1"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
    x="0px"
    y="0px"
    viewBox="0 0 278.454834 357.205414"
    style="enable-background:new 0 0 278.454834 357.205414;"
    xml:space="preserve"
    :src="src"
    @svg-loaded="loadSVG(...arguments)"
    @error="errorHappened(...arguments)"
    @show-part="showPart(...arguments)"
    class="avatar wrap"
  ></svg>
</template>

<script>
import _ from "lodash";
import * as axios from "axios";
import xml2js from "xml2js";
import adapter from "axios/lib/adapters/http";
export default {
    name: "Avatar",
    data: () => {
        return {
            src: "img/avatar1.svg",
            partsGrouped: undefined
        };
    },
    created() {
        this.$emit("svg-loaded", { src: "img/avatar1.svg" });
    },
    methods: {
        errorHappened(event) {
            console.error(event.err);
        },
        loadSVG: async function(event) {
            console.log("Event.options", event.src);
            // for some reason strings are not passed correctly
            let stringEvent = event.src;

            let src = stringEvent || this.$data.src;
            this.$data.src = src;
            console.info("SRC", src);
            await axios
                .get(`http://localhost:8080/${src}`, {
                    adapter
                })
                .then(response =>
                      xml2js
                      .parseStringPromise(response.data)
                      .catch(err => console.error("Error loading XML", err))
                      .then(
                          xml =>
                              (this.svg().innerHTML = new xml2js.Builder().buildObject(
                                  xml // TODO make sure we don't put nested SVG
                              ))
                      )
                      .then(() => {
                          console.log("partsGrouped!");
                          const parts = Array.prototype.slice.call(
                              this.svg().querySelectorAll("g[id]")
                          );
                          console.log("Parts:", parts);
                          this.showPart(
                              this.$parent.$parent.$refs.parts.$data.slides
                          );
                          this.$parent.$parent.$refs.parts.$forceUpdate();
                          return (this.$data.partsGrouped = _.groupBy(
                              parts,
                              it => it.id.split("_")[0]
                          ));
                      })
                     )
                .catch(err => this.$emit("errorHappened", { err: err }));
        },
        update() {
            _.each(this.$data.slides, (val, key) => {
                console.log("Key:Val", val, key);
                this.svg()
                    .rootElement.querySelectorAll("g[id *= '" + key + "_']")
                    .forEach(el => {
                        let id = el.getAttribute("id");
                        let visibility = id.endsWith("_" + (val + 1))
                            ? "visible"
                            : "hidden";
                        console.log("Setting", key, val + 1, id, "to", visibility);
                        el.setAttribute("visibility", visibility);
                    });
            });
        },

        /**
     Called from underlaying component when part should be shown, other hidden
    */
        showPart(slides) {
            console.log(slides);
            _.each(slides, (idx, name) => {
                console.log('idx:name', idx, name, this.svg().querySelectorAll('g[id]'));
                this.svg()
                    .querySelectorAll(`g[id][id *= '${name}_']`)
                    .forEach(el => {
                        console.log(el.id);
                        const visible = el.id.match(`_${idx + 1}$`) ? "visible" : "hidden";
                        el.setAttribute("visibility", visible);
                        console.log('idx:name', idx, name, visible);
                    })
          })
    }, // ,
    // changePart(name, idx) {
    //   _.each(this.$data.slides, (val, key) => {
    //     console.log("Key:Val", val, key);
    //     let svgDoc =
    //       document.querySelector("object").contentDocument ||
    //       document.querySelector("object").getSVGDocument();

    //     svgDoc.rootElement
    //       .querySelectorAll("g[id *= '" + key + "_']")
    //       .forEach(el => {
    //         let id = el.getAttribute("id");
    //         let visibility = id.endsWith("_" + (val + 1))
    //           ? "visible"
    //           : "hidden";
    //         console.log("Setting", key, val + 1, id, "to", visibility);
    //         el.setAttribute("visibility", visibility);
    //       });
    //   });
    // }
    svg() {
      return (
        this.$el.contentDocument ||
        (this.$el.getSVGDocument && this.$el.getSVGDocument()) ||
        this.$el
      );
    }
  }
};
</script>

<style></style>
