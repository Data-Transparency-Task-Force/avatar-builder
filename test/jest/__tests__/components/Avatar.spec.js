/* eslint-disable */
/**
 * @jest-environment jsdom
 */

import { mount, createLocalVue, shallowMount } from '@vue/test-utils'
import Avatar from '../../../../src/components/Avatar.vue'
import * as All from 'quasar'
import sinon from 'sinon'
import _ from "lodash";
// import langEn from 'quasar/lang/en-us' // change to any language you wish! => this breaks wallaby :(
const { Quasar, date } = All

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})

describe('Mount Quasar', () => {
  let localVue, wrapper, vm

  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(Quasar, { components }) // , lang: langEn

    wrapper = shallowMount(Avatar, {
      localVue
    })
    vm = wrapper.vm
  })

  jest.setTimeout(10000)
  it('passes the sanity check and creates a wrapper', () => {
    expect(wrapper.isVueInstance()).toBe(true)
  })
  describe("events", () => {
    it('`svg-loaded` event', () => {
      vm.$emit('svg-loaded', {src: '/img/some_img.svg'})
      wrapper.trigger('svg-loaded', {src: '/img/some_img.svg'})
      expect(wrapper.emitted()['svg-loaded'][1]).toEqual([{src:'/img/some_img.svg'}])
      expect(vm.$data.src).toBe("/img/some_img.svg")
    })

    it('`svg-loaded` even via stubs', () => {
      vm.loadSVG = sinon.stub()
      vm.errorHappened = sinon.stub()
      wrapper.trigger('svg-loaded', {src: '/img/some_img.svg'})
      expect(vm.loadSVG.called).toBe(true)
      expect(vm.errorHappened.called).toBe(false)
    })
  })

  describe("methods", () => {
    it("showPart sould properly set part visibility", async() => {
      expect(true).toBe(true)

      vm.errorHappened = sinon.stub()
      vm.showPart = sinon.stub()

      vm.$emit('svg-loaded', {src: 'img/avatar1.svg'})
      wrapper.trigger('svg-loaded', {src: "img/avatar1.svg"})

      expect(vm.$data.src).toBe("img/avatar1.svg")
      await new Promise(resolve => setTimeout(resolve, 1000))

      expect(vm.svg().querySelector('g[id]')).toBeDefined()
      expect(vm.svg().querySelector('g[id]')).not.toBeNull()
      expect(vm.svg().querySelectorAll('g[id]').length).toBe(24)


    })
    it("showPart sould properly set part visibility", async () => {
      console.log('---');

      vm.$emit('svg-loaded', {src: 'img/avatar2.svg'})
      wrapper.trigger('svg-loaded', {src: "img/avatar2.svg"})
      // letting component redraw
      await new Promise(resolve => setTimeout(resolve, 1000))
      console.log(vm.$data.src);
      expect(vm.$data.src).toBe("img/avatar2.svg")

      expect(vm.svg().querySelector('g[id]')).toBeDefined()
      expect(vm.svg().querySelector('g[id]')).not.toBeNull()
      expect(vm.svg().querySelectorAll('g[id]').length).toBe(58)


      const parts = vm.partsGrouped()
      const megaObject = _.reduce(parts, function(memo, current) { return _.assign(memo, current) },  {})
      expect(Object.keys(megaObject).length).toBe(10);

    })

    it("$event via stub", async () => {
      // checking loadSVG via stubs
      vm.loadSVG = sinon.stub()
      vm.errorHappened = sinon.stub()
      wrapper.trigger('svg-loaded', {src: "img/avatar2.svg"})
      expect(vm.loadSVG.called).toBe(true)
      expect(vm.loadSVG.callCount).toBe(1)
      expect(vm.errorHappened.called).toBe(false)
      //console.log(vm.partsGrouped())
    })
  })

  it('Snapshot matches', () => expect(vm.$el).toMatchSnapshot())

})
