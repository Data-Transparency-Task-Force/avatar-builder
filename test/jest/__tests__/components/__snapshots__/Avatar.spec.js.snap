// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount Quasar Snapshot matches 1`] = `
<svg
  class="avatar wrap"
  id="Layer_1"
  src="img/avatar1.svg"
  style="enable-background: new 0 0 278.454834 357.205414;"
  version="1.1"
  viewBox="0 0 278.454834 357.205414"
  x="0px"
  xml:space="preserve"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  y="0px"
/>
`;
